firmware/ethereum.c

```diff
@@ -0,0 +1,449 @@
+/*
+ * This file is part of the TREZOR project.
+ *
+ * Copyright (C) 2016 Alex Beregszaszi <alex@rtfs.hu>
+ * Copyright (C) 2016 Pavol Rusnak <stick@satoshilabs.com>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ethereum.h"
+#include "fsm.h"
+#include "layout2.h"
+#include "messages.h"
+#include "transaction.h"
+#include "ecdsa.h"
+#include "protect.h"
+#include "crypto.h"
+#include "secp256k1.h"
+#include "sha3.h"
+#include "util.h"
+
+static bool signing = false;
+static uint32_t data_total, data_left;
+static EthereumTxRequest resp;
+static uint8_t hash[32], sig[64], privkey[32];
+struct SHA3_CTX keccak_ctx;
+
+static inline void hash_data(const uint8_t *buf, size_t size)
+{
+	sha3_Update(&keccak_ctx, buf, size);
+}
+
+/*
+ * Push an RLP encoded length to the hash buffer.
+ */
+static void hash_rlp_length(uint32_t length, uint8_t firstbyte)
+{
+	uint8_t buf[4];
+	if (length == 1 && firstbyte == 0x00) {
+		// special case: null is encoded differently
```

Comment:
According to https://github.com/ethereum/wiki/wiki/RLP,  0x80 is the encoding of the empty string, not of the string containing a single 0 byte.

We shouldn't have the workaround here, instead I guess the higher layer should set value to the empty instead of "0", when a zero value  should be transferred.  Otherwise we cannot send a message with a single 0 byte.